cql/src/main/java/org/apache/cassandra/stargate/metrics/ClientMetrics.java

```diff
@@ -88,64 +102,114 @@ public void unpauseConnection() {
+   * @param meterRegistry Micrometer MeterRegistry to report to
+   * @param clientInfoTagProvider Tag provider for the connection
+   * @param updateRateMillis Number of milliseconds to schedule the internal metric update task. If
+   *     zero or less task will not be scheduled.
+   */
+  public synchronized void init(
+      Collection<Server> servers,
+      MeterRegistry meterRegistry,
+      ClientInfoMetricsTagProvider clientInfoTagProvider,
+      int updateRateMillis) {
+
     if (initialized) return;
 
     this.servers = servers;
-    this.metricRegistry = metricRegistry;
-
-    registerGauge("NettyDirectMemory", this::nettyDirectMemory);
-    registerGauge("NettyHeapMemory", this::nettyHeapMemory);
+    this.meterRegistry = meterRegistry;
+    this.clientInfoTagProvider = clientInfoTagProvider;
 
-    registerGauge("connectedNativeClients", this::countConnectedClients);
-    registerGauge("connectedNativeClientsByUser", this::countConnectedClientsByUser);
-    registerGauge("connections", this::connectedClients);
-    registerGauge("clientsByProtocolVersion", this::recentClientStats);
+    // netty gauges
+    meterRegistry.gauge(
+        metric("NettyDirectMemory"), Tags.empty(), this, ClientMetrics::nettyDirectMemory);
+    meterRegistry.gauge(
+        metric("NettyHeapMemory"), Tags.empty(), this, ClientMetrics::nettyHeapMemory);
 
-    authSuccess = registerMeter("AuthSuccess");
-    authFailure = registerMeter("AuthFailure");
-    authError = registerMeter("AuthError");
+    // connected native clients with the client info tags must be a multi gauge
+    connectedNativeClients =
+        MultiGauge.builder(metric("connectedNativeClients")).register(meterRegistry);
+    connectedNativeClientsByUser =
+        MultiGauge.builder(metric("connectedNativeClientsByUser")).register(meterRegistry);
 
-    pausedConnections = new AtomicInteger();
-    registerGauge("PausedConnections", pausedConnections::get);
-    requestDiscarded = registerMeter("RequestDiscarded");
-    requestsProcessed = registerMeter("RequestsProcessed");
+    pausedConnections =
+        meterRegistry.gauge(metric("PausedConnections"), Tags.empty(), new AtomicInteger(0));
 
-    totalBytesRead = registerCounter("TotalBytesRead");
-    totalBytesWritten = registerCounter("TotalBytesWritten");
+    totalBytesRead = meterRegistry.counter(metric("TotalBytesRead"));
+    totalBytesWritten = meterRegistry.counter(metric("TotalBytesWritten"));
 
-    bytesReceivedPerFrame = registerHistogram("BytesReceivedPerFrame");
-    bytesTransmittedPerFrame = registerHistogram("BytesTransmittedPerFrame");
+    bytesReceivedPerFrame = meterRegistry.summary(metric("BytesReceivedPerFrame"));
+    bytesTransmittedPerFrame = meterRegistry.summary(metric("BytesTransmittedPerFrame"));
 
     initialized = true;
+
+    // if we have the update rate, init the executor service and submit the update task
+    if (updateRateMillis > 0) {
```

Comment:
Why do we need this if? If someone sets it to `0`, shouldn't the below action just execute immediately?

Reply:
This is there so somebody can disable the update with the value zero. Since we make it configurable rate, we should also make it configurable to be turned off.

Reply:
Maybe we should consider adding dedicated on/off settings instead of relying on `updateRateMillis ==0`?